metadata-io/src/main/java/com/linkedin/metadata/entity/RetentionService.java

```diff
@@ -0,0 +1,207 @@
+   *                   meaning it will be the default for any entities without specified policy
+   * @param aspectName Aspect name to apply policy to. If null, set as "*",
+   *                   meaning it will be the default for any aspects without specified policy
+   * @param retentionInfo Retention policy
+   */
+  @SneakyThrows
+  public boolean setRetention(@Nullable String entityName, @Nullable String aspectName,
+      @Nonnull DataHubRetentionInfo retentionInfo) {
+    validateRetentionInfo(retentionInfo);
+    DataHubRetentionKey retentionKey = new DataHubRetentionKey();
+    retentionKey.setEntityName(entityName != null ? entityName : ALL);
+    retentionKey.setAspectName(aspectName != null ? aspectName : ALL);
+    Urn retentionUrn = EntityKeyUtils.convertEntityKeyToUrn(retentionKey, DATAHUB_RETENTION_ENTITY);
+    MetadataChangeProposal keyProposal = new MetadataChangeProposal();
+    GenericAspect keyAspect = GenericAspectUtils.serializeAspect(retentionKey);
+    keyProposal.setAspect(keyAspect);
+    keyProposal.setAspectName(DATAHUB_RETENTION_KEY_ASPECT);
+    keyProposal.setEntityType(DATAHUB_RETENTION_ENTITY);
+    keyProposal.setChangeType(ChangeType.UPSERT);
+    keyProposal.setEntityUrn(retentionUrn);
+    AuditStamp auditStamp =
+        new AuditStamp().setActor(Urn.createFromString(Constants.SYSTEM_ACTOR)).setTime(System.currentTimeMillis());
+    getEntityService().ingestProposal(keyProposal, auditStamp);
+    MetadataChangeProposal aspectProposal = keyProposal.clone();
+    GenericAspect retentionAspect = GenericAspectUtils.serializeAspect(retentionInfo);
+    aspectProposal.setAspect(retentionAspect);
+    aspectProposal.setAspectName(DATAHUB_RETENTION_ASPECT);
+    return getEntityService().ingestProposal(aspectProposal, auditStamp).isDidUpdate();
+  }
+
+  /**
+   * Delete the retention policy set for given entity and aspect.
+   *
+   * @param entityName Entity name to apply policy to. If null, set as "*",
+   *                   meaning it will delete the default policy for any entities without specified policy
+   * @param aspectName Aspect name to apply policy to. If null, set as "*",
+   *                   meaning it will delete the default policy for any aspects without specified policy
+   */
+  public void deleteRetention(@Nullable String entityName, @Nullable String aspectName) {
+    DataHubRetentionKey retentionKey = new DataHubRetentionKey();
+    retentionKey.setEntityName(entityName != null ? entityName : ALL);
+    retentionKey.setAspectName(aspectName != null ? aspectName : ALL);
+    Urn retentionUrn = EntityKeyUtils.convertEntityKeyToUrn(retentionKey, DATAHUB_RETENTION_ENTITY);
+    getEntityService().deleteUrn(retentionUrn);
+  }
+
+  private void validateRetentionInfo(DataHubRetentionInfo retentionInfo) {
+    Set<String> retentionsSoFar = new HashSet<>();
+    for (Retention retention : retentionInfo.getRetentionPolicies()) {
+      if (retention.data().size() != 1) {
```

Comment:
i thought we were supporting either or?

Reply:
yup so we support it as list of retention. Inside a single Retention object, we should have just one set, kind of emulating a union behavior. We could instead have a single Retention object instead of a list of retention objects as policy and set as many policies as needed. WDYT?